data/padics/padic_norm.lean

```diff
@@ -0,0 +1,448 @@
+           (padic_val (prime.gt_one p_prime) (q.denom*r.num)),
+    from calc 
+    min ((padic_val (prime.gt_one p_prime) (q.num*r.denom)) : ℤ) 
+           (padic_val (prime.gt_one p_prime) (q.denom*r.num)) = 
+    (min (padic_val (prime.gt_one p_prime) (q.num*r.denom)) 
+           (padic_val (prime.gt_one p_prime) (q.denom*r.num)) : ℤ) :
+       by simp
+     ... ≤ (padic_val (prime.gt_one p_prime) (q.num * r.denom + q.denom * r.num) : ℤ) : 
+       begin 
+         apply min_le_iff.2, 
+         simp only [int.coe_nat_le], 
+         apply min_le_iff.1, 
+         apply min_le_padic_val_add, 
+         assumption 
+       end,  
+calc 
+  padic_val_rat (prime.gt_one p_prime) (q + r) 
+         = padic_val_rat (prime.gt_one p_prime) 
+                         (((q.num * r.denom + q.denom * r.num : ℤ)) /. (↑q.denom * ↑r.denom : ℤ)) :
+            by rw hqreq
+     ... = padic_val (prime.gt_one p_prime) 
+                     (q.num * r.denom + q.denom * r.num) - 
+            padic_val (prime.gt_one p_prime) (↑q.denom * ↑r.denom) :
+       begin apply padic_val_rat.defn, cc, reflexivity end 
+     ... ≥ min (padic_val (prime.gt_one p_prime) (q.num*r.denom)) 
+               (padic_val (prime.gt_one p_prime) (q.denom*r.num)) - 
+            padic_val (prime.gt_one p_prime) (↑q.denom * ↑r.denom) :
+       sub_le_sub hge (le_refl _)
+     ... = min (padic_val (prime.gt_one p_prime) q.num + 
+                 padic_val (prime.gt_one p_prime) r.denom) 
+               (padic_val (prime.gt_one p_prime) q.denom + 
+                 padic_val (prime.gt_one p_prime) r.num) - 
+            (padic_val (prime.gt_one p_prime) q.denom + 
+              padic_val (prime.gt_one p_prime) r.denom) :
+       begin rw [←padic_val.mul, ←padic_val.mul, ←padic_val.mul], reflexivity, repeat {simpa} end
+     ... = min (padic_val (prime.gt_one p_prime) q.num + 
+                 padic_val (prime.gt_one p_prime) r.denom - 
+                 (padic_val (prime.gt_one p_prime) q.denom + 
+                   padic_val (prime.gt_one p_prime) r.denom)) 
+               (padic_val (prime.gt_one p_prime) q.denom + 
+                 padic_val (prime.gt_one p_prime) r.num - 
+                 (padic_val (prime.gt_one p_prime) q.denom + 
+                   padic_val (prime.gt_one p_prime) r.denom)) :
+       by apply min_add_eq_min_add_min
+     ... = min (padic_val (prime.gt_one p_prime) q.num - 
+                 padic_val (prime.gt_one p_prime) q.denom) 
+               (padic_val (prime.gt_one p_prime) r.num - 
+                 padic_val (prime.gt_one p_prime) r.denom) :
+       by simp
+     ... = min (padic_val_rat (prime.gt_one p_prime) (q.num /. ↑q.denom)) 
+               (padic_val_rat (prime.gt_one p_prime) (r.num /. ↑r.denom)) : 
+       by rw [padic_val_rat.defn, padic_val_rat.defn]; simpa
+     ... = min (padic_val_rat (prime.gt_one p_prime) q) 
+               (padic_val_rat (prime.gt_one p_prime) r) : 
+       by rw [←rat.num_denom q, ←rat.num_denom r]
```

Comment:
warning: I will probably rewrite this

Reply:
Please. I thought about rewriting this today as I was cleaning things up, but got scared.